### PR TITLE
Add demo CoAP for IoT Core

### PR DESCRIPTION
**Pull request by ptone**
To test from staged location:

git clone https://github.com/ptone/community.git
cd community
git checkout coap-proxy
cd tutorials/cloud-iot-coap-proxy

### CODE REVIEW

File: tutorials/cloud-iot-coap-proxy/proxy/src/main/java/com/agosto/iot/IotCoreForwardingResource.java [hunk 1]
```diff
@@ -0,0 +1,247 @@
+package com.agosto.iot;
```
Comment (gguuss) [inline tutorials/cloud-iot-coap-proxy/proxy/src/main/java/com/agosto/iot/IotCoreForwardingResource.java]: License header on this file is missing

File: tutorials/cloud-iot-coap-proxy/proxy/src/test/java/com/agosto/iot/AppTest.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package com.agosto.iot;
```
Comment (gguuss) [inline tutorials/cloud-iot-coap-proxy/proxy/src/test/java/com/agosto/iot/AppTest.java]: What is this testing?

File: tutorials/cloud-iot-coap-proxy/coap-dtls-client/client.env [hunk 3]
```diff
@@ -0,0 +1,26 @@
+#! /bin/bash
```
Comment (gguuss) [inline tutorials/cloud-iot-coap-proxy/coap-dtls-client/client.env]: consider adding a copyright / license header to your shell scripts and other code that doesn't have a header

**Comment by gguuss** [comment]
Assigning to Todd, I just had nits, approving.

**Comment by ToddKopriva** [comment]
I'll do the editorial pass soon, probably Monday (4Feb).

**Comment by gguuss** [comment]
@ptone PTAL on my comments too, mostly nits, may not require changes.

File: tutorials/cloud-iot-coap-proxy/coap-dtls-client/src/main/java/com/agosto/coap/dtls/client/JwtBuilder.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
+package com.agosto.coap.dtls.client;
```
[the remark this answers is not in the document]
Comment (ptone) [reply tutorials/cloud-iot-coap-proxy/coap-dtls-client/src/main/java/com/agosto/coap/dtls/client/JwtBuilder.java]: This was the partner's choice

File: tutorials/cloud-iot-coap-proxy/coap-dtls-client/dependency-reduced-pom.xml [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  <groupId>com.agosto.coap.dtls.client</groupId>
+  <artifactId>coap-dtls-client</artifactId>
+  <name>coap-dtls-client</name>
+  <version>1.0-SNAPSHOT</version>
+  <url>http://www.example.com</url>
+  <build>
+    <plugins>
+      <plugin>
+        <artifactId>maven-surefire-plugin</artifactId>
+        <version>3.0.0-M1</version>
```
[the remark this answers is not in the document]
Comment (ptone) [reply tutorials/cloud-iot-coap-proxy/coap-dtls-client/dependency-reduced-pom.xml]: fixed

File: tutorials/cloud-iot-coap-proxy/coap-dtls-client/client.env (hunk 3 above)
Comment (ptone) [reply tutorials/cloud-iot-coap-proxy/coap-dtls-client/client.env]: FWIW - the community writing guidelines do not prescribe that (maybe it should?) - I'll do that for googler authored stuff in general, this was contractor code as we discussed in chat - I'm going to consider this is covered by CLA

**Comment by ptone** [comment]
addressed comments, chatted with Gus and merged latest from master @ToddKopriva ready for your final process steps - thx

**Comment by ToddKopriva** [comment]
I'll have a look soon.

**Comment by boaks** [comment]
That's great!

Are there any reasons, why you use the "old" 2.0.0-M11 ? Have you been faced any issues using M12 or M13?

Just for the case,  you're not aware of it:
- Californium/Scandium runs also on Android. (If you're interested, I can push my "not that complete elaborated" extended android client to the californium repo.)
- The upcoming M14 includes the current DTLS 1.2 CID extension (see
https://github.com/tlswg/dtls-conn-id), which overcomes not only the NAT, you can also change from WLAN to Cellular and back without handshakes.

The DTLS 1.2 CID draft is not finalized yet, and the redesigned DTLS layer in M14 may currently need some more work, but at least it works in way, that you can imagine, what the benefits will be.

**Comment by ptone** [comment]
@boaks This was built a bit ago, as part of some customer/partner POC work and we just wanted to get it out there as an illustration point.  PRs are welcome if you would like to update it to be more current.  There are other aspects where this could be improved from a production-readiness POV in terms of running it in GKE or as an instance group.

**Comment by boaks** [comment]
I managed to actually try it out. Really nice! 

My next steps:

> PRs are welcome if you would like to update it to be more current.

I will check with the open-source-management of my employer, if I can go for corporate CLA, or a individual CLA. May be, there is already a corporate-relation in our open-source-management, then it should be possible with in couple of days. If not, it may take longer ...

I will spend some more time next weekend, to check, if the upcoming californium 2.0.0-M14 could be used just by updating the dependency. Or, if not, modify it a little more ...
(Though the M14 is mainly request by a californium user, which has done very valuable testing for the californium project, I have ask him, if postpone the milestone for some day will be OK for him.)

> This was built a bit ago, as part of some customer/partner POC work

Is there more public information available the the POC targets and results?

> There are other aspects where this could be improved from a production-readiness POV in terms of running it in GKE or as an instance group.

Clustering dtls mainly suffers from missing "load-balancer-support". Without that support, the clients are forced to frequently execute a dtls handshake. That somehow thwarts the benefits of  dtls with connection id, where the handshakes could be reduce to a minimum.
I would recommend, to only check, that it is "principally working", but then go with a single instance. For the most IoT use-cases that's quite good enough for a "long time".

**Comment by boaks** [comment]
I checked the tutorial with the new californium milestone release 2.0.0-M14.
Beside of the change in the `CoapEndpoint`, the tutorial works with that new 2.0.0-M14.

The `CoapEndpoint.CoapEndpointBuilder` was renamed to `CoapEndpoint.Builder`, which affects the cloud based `coap-iot-core`. We also removed the already deprecated `CoapEndpoint` constructors and made the use of the `CoapEndpoint.Builder` mandatory. That affects the `coap-dtls-client` to create the `CoapEndpoint` similar to the cloud part.

> corporate CLA
The good part of the news is, that I wasn't the first to ask for that. The downside is unfortunately, that the first check for that corporate CLA some month ago results in "major concerns by central legal". That's the downside of working for one of the big engineer and production companies in Germany. We will see. The current forecast is something about a couple months for a decisions.   

> individual CLA
The combination of "major concerns" and being a committer in that "close related project for this contribution" results in a kindly request of my open source officer not to go for that.

In sum, it will take some time to either get the `corporate CLA` on the way or to reask for the permission of the `individual CLA`.

**Pull request merged**